### PR TITLE
wsContainersAttach attach to stdin/out/err streams as requested

### LINKED ISSUE #43272: Attaching websocket (/containers/{name:.*}/attach/ws) not configurable as stated in the docs
**Description**

The [docs](https://docs.docker.com/engine/api/v1.41/#operation/ContainerAttachWebsocket) describe, that `ContainerAttachWebsocket`  is configurable for stdin/stdout/stderr, but stdin and stderr are always merged, independently of the actual params.

Possible, the reason is here:
If I followed the code correctly, 
https://github.com/moby/moby/blob/7b9275c0da707b030e62c96b679a976f31f929d3/api/server/router/container/container.go#L45
links to the function wsContainersAttach
https://github.com/moby/moby/blob/9e7bbdb9ba53440cfc67e2e1de89e35e44c43ddf/api/server/router/container/container_routes.go#L640
which sets stdin, stdout and stderr to true and actually ignores the params:
https://github.com/moby/moby/blob/9e7bbdb9ba53440cfc67e2e1de89e35e44c43ddf/api/server/router/container/container_routes.go#L676-L685

**Steps to reproduce the issue:**

Just have a container with some stdout:
```bash
docker run ubuntu bash -c "echo 'Hello World' && sleep 1000"
```

Connecting to the container using python docker api:
```py
import docker

api_client = docker.APIClient('yourdockersocket')
# Uses /containers/{name:.*}/attach/ws (Websocket Stream)
sock = api_client.attach_socket('test_container_id', params={'stderr': 1, 'stream': 1, 'logs': 1}, ws=True)
print(sock1.read(11))
```
will output stdout as well, although only stderr is desired. On the other hand:
```py
import docker

api_client = docker.APIClient('yourdockersocket')
# Uses /containers/{name:.*}/attach (TCP stream)
sock = api_client.attach_socket('test_container_id', params={'stderr': 1, 'stream': 1, 'logs': 1})
print(sock1.read(11))
```
will correctly, probably using
https://github.com/moby/moby/blob/9e7bbdb9ba53440cfc67e2e1de89e35e44c43ddf/api/server/router/container/container_routes.go#L613-L622

**Describe the results you received:**
As explained above, stderr and stdout merged stream is returned in the websocket endpoint, although not specified.

**Describe the results you expected:**
As explained above, expected result would be the stderr stream and stdout stream separately, but they are merged for the websocket request.

**Output of `docker version`:**

```
Client: Docker Engine - Community
 Version:           20.10.12
 API version:       1.41
 Go version:        go1.16.12
 Git commit:        e91ed57
 Built:             Mon Dec 13 11:45:33 2021
 OS/Arch:           linux/amd64
 Context:           default
 Experimental:      true

Server: Docker Engine - Community
 Engine:
  Version:          20.10.12
  API version:      1.41 (minimum version 1.12)
  Go version:       go1.16.12
  Git commit:       459d0df
  Built:            Mon Dec 13 11:43:42 2021
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.4.12
  GitCommit:        7b11cfaabd73bb80907dd23182b9347b4245eb5d
 runc:
  Version:          1.0.2
  GitCommit:        v1.0.2-0-g52b36a2
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0
```

**Output of `docker info`:**

```
Client:
 Context:    default
 Debug Mode: false
 Plugins:
  app: Docker App (Docker Inc., v0.9.1-beta3)
  buildx: Docker Buildx (Docker Inc., v0.7.1-docker)
  scan: Docker Scan (Docker Inc., v0.12.0)

Server:
 Containers: 1
  Running: 1
  Paused: 0
  Stopped: 0
 Images: 25
 Server Version: 20.10.12
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Native Overlay Diff: true
  userxattr: false
 Logging Driver: json-file
 Cgroup Driver: cgroupfs
 Cgroup Version: 1
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: io.containerd.runc.v2 io.containerd.runtime.v1.linux runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 7b11cfaabd73bb80907dd23182b9347b4245eb5d
 runc version: v1.0.2-0-g52b36a2
 init version: de40ad0
 Security Options:
  seccomp
   Profile: default
 Kernel Version: 5.10.60.1-microsoft-standard-WSL2
 Operating System: Ubuntu 20.04.3 LTS
 OSType: linux
 Architecture: x86_64
 CPUs: 8
 Total Memory: 12.32GiB
 Name: DEAGM03731
 ID: 4GVK:G7FV:GK5N:W5KE:2X7M:44FA:26XI:B5MW:DTLG:NLQV:7XLV:FK5L
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 Registry: https://index.docker.io/v1/
 Labels:
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Registry Mirrors:
  https://docker.deagxartifactory.sickcn.net/
 Live Restore Enabled: false

WARNING: API is accessible on http://127.0.0.1:2375 without encryption.
         Access to the remote API is equivalent to root access on the host. Refer
         to the 'Docker daemon attack surface' section in the documentation for
         more information: https://docs.docker.com/go/attack-surface/
WARNING: No blkio throttle.read_bps_device support
WARNING: No blkio throttle.write_bps_device support
WARNING: No blkio throttle.read_iops_device support
WARNING: No blkio throttle.write_iops_device support
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**
Running on WSL2.

### PR DESCRIPTION
**Pull request by ndeloof**
**- What I did**
`attach/ws` API applies stdlin/stdout/stderr parameters to select streams to attach to (as documented)
closes #43272

**- How I did it**
set parameters on `ContainerAttachConfig`. Also aligned struct initialization with `postContainersAttach` for consistency.

**- How to verify it**

curl -v -H 'Connection: upgrade' -H 'upgrade: websocket' -H 'Sec-Websocket-Key:hello' -H 'Sec-Websocket-Version: 13'tp://localhost:2375/v1.41/containers/$ID/attach/ws?stream=1&stdout=1"  --output -

**- Description for the changelog**
`containers/{ID}/attach/ws` only attach to streams according by `stdin`, `stdout` and `stderr` parameters.

**- A picture of a cute animal (not mandatory but encouraged)**
image

### CODE REVIEW

**Review by cpuguy83 — COMMENTED** [review]
While the change is the correct behavior, it breaks older clients completely.

I put the TODO in there originally because it had already been years of the existing broken behavior.
I'd also expect us to take advantage of ws channels rather than doing the muxing on the data.

**Comment by ndeloof** [comment]
@cpuguy83 I replaced TODO with a comment, but still multiplex is set to `false`, so we don't use data multiplexing, but just plain websocket transport
The only change here is to allow clients to select the streams to be attached, so they can only attach to stdout and ignore stderr (for sample)

**Comment by cpuguy83** [comment]
Thanks for clarifying, I see that now. Don't know why I didn't see that before.

**Comment by cpuguy83** [comment]
I think we should keep the existing behavior for old API versions.

**Comment by thaJeztah** [comment]
Yes, if this breaks current clients, we should probably make the (default) behavior depend on the API version. (Haven't looked in depth what/how things break)

Perhaps would be good to;

- describe the change in the commit message (perhaps refer to the commit that changed it, that I mentioned in https://github.com/moby/moby/issues/43272#issuecomment-1048950945 - if that's indeed the commit that broke it); FWIW; I try to avoid mentioning issue numbers in the commit message, as they can generate a lot of "noise" on github, and usually refer to _commit_ in the commit message itself - this also allows finding the change back locally (without github))
- I haven't checked what the API documentation describes currently (swagger); perhaps you can check if the documentation describes it correctly.
- same for the API version-history; https://github.com/moby/moby/blob/master/docs/api/version-history.md

(Happy to help with  rephrasing / wording if needed)

**Comment by ndeloof** [comment]
I've checked endpoint history, and it actually never supported streams to be selected. Initial implementation used a distinct backend function for HTTP and WS attachment, which was simplified into a single one by https://github.com/moby/moby/commit/a77b7dd2278106b9081d0ef2260fbeea790a91ef.
I've updated this PR accordingly:
- parameters are only used to select streams if API >= 1.42
- API change is documented
- swagger is already correct

**Review by cpuguy83 — APPROVED** [review]

**Comment by thaJeztah** [comment]
> and it actually never supported streams to be selected. Initial implementation used a distinct backend function for HTTP and WS attachment, which was simplified into a single one by https://github.com/moby/moby/commit/a77b7dd2278106b9081d0ef2260fbeea790a91ef.

Interesting; so it didn't work before https://github.com/moby/moby/commit/c44f513248a8b40b1b2221726c7441881383e919 (https://github.com/moby/moby/pull/12120) either? https://github.com/moby/moby/issues/43272#issuecomment-1048950945

**Comment by thaJeztah** [comment]
So, looks like the documentation for existing API versions is incorrect, as it mentions the `default` to be `false`; https://docs.docker.com/engine/api/v1.41/#operation/ContainerAttachWebsocket

Screenshot 2022-05-18 at 10 39 09

File: api/server/router/container/container_routes.go [hunk 1]
```diff
@@ -693,15 +693,22 @@ func (s *containerRouter) wsContainersAttach(ctx context.Context, w http.Respons
 		return conn, conn, conn, nil
 	}
 
+	useStdin, useStdout, useStderr := true, true, true
+	if versions.GreaterThanOrEqualTo(version, "1.42") {
+		useStdin = httputils.BoolValue(r, "stdin")
+		useStdout = httputils.BoolValue(r, "stdout")
+		useStderr = httputils.BoolValue(r, "stderr")
+	}
```
Comment (thaJeztah) [inline api/server/router/container/container_routes.go:701]: Trying to think if there's a way to make the change less disruptive, but not sure how to;

- API < 1.41 enabled these by default (which didn't match the docs (see my other comment)), so I'd consider that to be a bug
- But has been around for a long time, so we probably shouldn't change for existing API versions

So in the old API versions; if I didn't look at the API docs, and forgot to set `?stdin=1&stdout=1&stderr=1`, then I would _still_ have all three attached.

In the new API version; I _MUST_ set all three to preserve the old behavior; which matches what was documented, but (see above) "just worked" before.


So, yeah, I don't see a clear way to migrate between one and the other, other than calling it out explicitly in the release notes 🤔 🤷‍♂️

File: docs/api/version-history.md [hunk 2]
```diff
@@ -87,6 +87,8 @@ keywords: "API, Docker, rcli, REST, documentation"
 * The `Volume` type, as returned by `Added new `ClusterVolume` fields 
 * Added a new `PUT /volumes{name}` endpoint to update cluster volumes (CNI).
   Cluster volumes are only supported if the daemon is a Swarm manager.
+* `/containers/{name}/attach/ws` endpoint only attach to configured streams
```
Comment (thaJeztah) [inline docs/api/version-history.md:90]: nit;

```suggestion
* `GET /containers/{name}/attach/ws` endpoint only attach to configured streams
```

(we can fix that later, not a blocker)

**Review by thaJeztah — APPROVED** [review]
LGTM

**Pull request merged**